File: src/aion/api/nn/mlp.zig
// SPDX-License-Identifier: EPL-2.0 OR GPL-2.0-or-later
const std = @import("std");

const layer_mod = @import("layer.zig");
const state_mod = @import("state.zig");
const linear_mod = @import("linear.zig");
const activation = @import("activation.zig");

pub const Builder = layer_mod.Builder;
pub const TensorRef = layer_mod.TensorRef;
pub const Tensor = layer_mod.Tensor;
pub const LayerName = layer_mod.LayerName;
pub const Params = state_mod.Params;
pub const BindError = state_mod.BindError;
pub const Linear = linear_mod.Linear;
pub const Kind = activation.Kind;

/// `w2(act(w1(x)))` — the plain two-matmul feed-forward block.
pub const FeedForward = struct {
    fc1: Linear,
    fc2: Linear,
    act: Kind = .silu,
    id: LayerName = .{},

    const Self = @This();

    pub const Weights = struct {
        fc1: Linear.Weights,
        fc2: Linear.Weights,
    };

    pub const Options = struct {
        /// Scope segment for this block, prefixing both projections.
        name: ?[]const u8 = null,
        act: Kind = .silu,
    };

    pub fn bind(bld: *Builder, params: anytype, opts: Options) BindError!Self {
        var src = state_mod.binding(Weights, params);
        const p: Params = src.params();

        const id, const scope = try LayerName.open(Self, bld, p.name orelse opts.name);
        defer bld.endScope(scope);

        return .{
            .fc1 = try Linear.bind(bld, p.child(Weights, .fc1), .{}),
            .fc2 = try Linear.bind(bld, p.child(Weights, .fc2), .{}),
            .act = opts.act,
            .id = id,
        };
    }

    pub fn forward(self: Self, bld: *Builder, x: TensorRef) Builder.Error!TensorRef {
        const scope = try self.id.enter(bld);
        defer bld.endScope(scope);

        const h: TensorRef = try self.fc1.forward(bld, x);
        const a: TensorRef = try activation.apply(bld, self.act, h);
        return self.fc2.forward(bld, a);
    }
};

/// `down(act(gate(x)) * up(x))` — the gated feed-forward used by SwiGLU (`.silu`)
/// and GeGLU (`.gelu`) transformers.
///
/// `gate` and `up` are two separate projections, which is how every checkpoint
/// ships them. Pre-concatenating the pair into one `[in, 2*ffn]` weight is a
/// *fusion*, not a layout, and it is the compiler's: `opt/fuse_horizontal_matmul`
/// rewrites matmuls sharing an operand into one wide matmul plus a slice each,
/// numerically identically. So there is nothing to choose here and nothing to
/// probe for — one shape, one code path.
///
/// The gelu case routes through the fused `geluMul` op rather than a separate
/// unary and multiply.
pub const GatedMLP = struct {
    gate: Linear,
    up: Linear,
    down: Linear,
    act: Kind,
    id: LayerName = .{},

    const Self = @This();

    pub const Weights = struct {
        gate_proj: Linear.Weights,
        up_proj: Linear.Weights,
        down_proj: Linear.Weights,
    };

    pub const Options = struct {
        /// Scope segment for this block, prefixing its projections.
        name: ?[]const u8 = null,
        act: Kind = .silu,
    };

    pub fn bind(bld: *Builder, params: anytype, opts: Options) BindError!Self {
        var src = state_mod.binding(Weights, params);
        const p: Params = src.params();

        const id, const scope = try LayerName.open(Self, bld, p.name orelse opts.name);
        defer bld.endScope(scope);

        const gate = try Linear.bind(bld, p.child(Weights, .gate_proj), .{});
        const up = try Linear.bind(bld, p.child(Weights, .up_proj), .{});
        const down = try Linear.bind(bld, p.child(Weights, .down_proj), .{});

        // The two halves are multiplied elementwise, so a width mismatch is a wiring
        // error that would otherwise surface as a confusing shape failure downstream.
        const ffn = lastDim(bld, gate.w) orelse return error.InvalidArgument;
        if (ffn == 0 or (lastDim(bld, up.w) orelse 0) != ffn) return error.InvalidArgument;

        return .{ .gate = gate, .up = up, .down = down, .act = opts.act, .id = id };
    }

    pub fn forward(self: Self, bld: *Builder, x: TensorRef) Builder.Error!TensorRef {
        const scope = try self.id.enter(bld);
        defer bld.endScope(scope);

        const gate_v: TensorRef = try self.gate.forward(bld, x);
        const up_v: TensorRef = try self.up.forward(bld, x);

        // `geluMul` fuses gelu(a)*b; the converters hand-rolled the pair and missed
        // this op entirely.
        const gated: TensorRef = if (self.act == .gelu)
            try bld.geluMul(gate_v, up_v)
        else
            try bld.mul(try activation.apply(bld, self.act, gate_v), up_v);

        return self.down.forward(bld, gated);
    }
};

/// `a * sigmoid(b)` over a projection split in half along the last dim — the GLU
/// in a Conformer convolution module.
pub const GLU = struct {
    proj: Linear,
    half: usize,
    id: LayerName = .{},

    const Self = @This();

    pub const Weights = struct {
        proj: Linear.Weights,
    };

    pub const Options = struct {
        name: ?[]const u8 = null,
    };

    pub fn bind(bld: *Builder, params: anytype, opts: Options) BindError!Self {
        var src = state_mod.binding(Weights, params);
        const p: Params = src.params();

        const id, const scope = try LayerName.open(Self, bld, p.name orelse opts.name);
        defer bld.endScope(scope);

        const proj = try Linear.bind(bld, p.child(Weights, .proj), .{});
        const total: usize = lastDim(bld, proj.w) orelse return error.InvalidArgument;
        if (total == 0 or total % 2 != 0) return error.InvalidArgument;

        return .{ .proj = proj, .half = total / 2, .id = id };
    }

    pub fn forward(self: Self, bld: *Builder, x: TensorRef) Builder.Error!TensorRef {
        const scope = try self.id.enter(bld);
        defer bld.endScope(scope);

        const both: TensorRef = try self.proj.forward(bld, x);
        const a: TensorRef = try bld.sliceLastDim(both, 0, self.half);
        const b: TensorRef = try bld.sliceLastDim(both, self.half, self.half);
        return bld.mul(a, try bld.sigmoid(b));
    }
};

fn lastDim(bld: *Builder, ref: TensorRef) ?usize {
    const shape = bld.knownShape(ref) orelse return null;
    if (shape.len == 0) return null;
    return shape[shape.len - 1];
}

File: bindings/python/src/aion/nn/mlp.py
# SPDX-License-Identifier: Apache-2.0
"""Feed-forward blocks, mirroring `src/aion/api/nn/mlp.zig`."""
from __future__ import annotations

from typing import Optional

from ..types import DTypeLike
from ..builder import TensorRef, WeightData
from .layers import Linear
from .module import Module, builder_of


class FeedForward(Module):
    """`w2(act(w1(x)))` — the plain two-matmul feed-forward block."""

    def __init__(
        self,
        w1: WeightData,
        w2: WeightData,
        *,
        act: str = "silu",
        name: Optional[str] = None,
        bias1: Optional[WeightData] = None,
        bias2: Optional[WeightData] = None,
        dtype: DTypeLike = "f32",
    ) -> None:
        self._layer_name = name
        self.act = act
        self.fc1 = Linear(w1, bias1, dtype=dtype)
        self.fc2 = Linear(w2, bias2, dtype=dtype)

    def forward(self, x: TensorRef) -> TensorRef:
        b = builder_of(x)
        with self._scoped(b):
            return self.fc2(b.unary(self.act, self.fc1(x)))


class GatedMLP(Module):
    """`down(act(gate(x)) * up(x))` — SwiGLU (`silu`) / GeGLU (`gelu`).

    `gate` and `up` are two separate projections, the way every checkpoint ships
    them. Pre-concatenating the pair into one `[in, 2*ffn]` weight is a *fusion*,
    not a layout, and it belongs to the compiler: `opt/fuse_horizontal_matmul`
    rewrites matmuls sharing an operand into one wide matmul plus a slice each,
    numerically identically. So there is nothing to choose here.

    The gelu case routes through the fused `gelu_mul` op rather than a separate
    unary and multiply.
    """

    def __init__(
        self,
        gate: WeightData,
        up: WeightData,
        down: WeightData,
        *,
        act: str = "silu",
        name: Optional[str] = None,
        dtype: DTypeLike = "f32",
    ) -> None:
        self._layer_name = name
        self.act = act

        ffn = int(_last_dim(gate))
        if int(_last_dim(up)) != ffn:
            raise ValueError("gate and up must have the same output width")

        self.gate_proj = Linear(gate, dtype=dtype)
        self.up_proj = Linear(up, dtype=dtype)
        self.down_proj = Linear(down, dtype=dtype)

    def forward(self, x: TensorRef) -> TensorRef:
        b = builder_of(x)
        with self._scoped(b):
            g = self.gate_proj(x)
            u = self.up_proj(x)
            gated = b.gelu_mul(g, u) if self.act == "gelu" else b.mul(b.unary(self.act, g), u)
            return self.down_proj(gated)


class GLU(Module):
    """`a * sigmoid(b)` over a projection split in half along the last dim."""

    def __init__(
        self,
        weight: WeightData,
        bias: Optional[WeightData] = None,
        *,
        name: Optional[str] = None,
        dtype: DTypeLike = "f32",
    ) -> None:
        self._layer_name = name
        total = int(_last_dim(weight))
        if total % 2 != 0:
            raise ValueError(f"GLU projection width must be even, got {total}")
        self.half = total // 2
        self.proj = Linear(weight, bias, dtype=dtype)

    def forward(self, x: TensorRef) -> TensorRef:
        b = builder_of(x)
        with self._scoped(b):
            both = self.proj(x)
            a = b.slice_last_dim(both, 0, self.half)
            g = b.slice_last_dim(both, self.half, self.half)
            return b.mul(a, b.unary("sigmoid", g))


def _last_dim(data: object) -> int:
    from ..builder import _infer_shape

    shape = _infer_shape(data)
    if not shape:
        raise ValueError("weight must have at least one axis")
    return int(shape[-1])
